### PR TITLE
[RFC] treewide: tidy up image recipes

### PR DESCRIPTION
**Pull request by mans0n**
- clean up `check-size $$$$(IMAGE_SIZE)`
- clean up `DEVICE_VARS`
- replace `${}` with `$()`
- replace `$$$$()` with `$$()`

### CODE REVIEW

**Comment by mans0n** [comment]
My original goal was the last one, replacing `$$$$()` with `$$()`, but I'm concerned for its side effects. I wrote a sample patch for apm821xx. Depending on the feedback, I'll either write up treewide patch, or I'll just drop it.

**Comment by adrianschmutzler** [comment]
Hi, I plan to relatively quickly merge patches 1, 4, 6 and 7 (in case someone wants to complain).

For no. 8, please replace the "if certain conditions are met." with what those conditions would be, as that's exactly the question relevant for understanding/rating this patch.

**Comment by adrianschmutzler** [comment]
patch 3: "Now that check-size uses IMAGE_SIZE by default, skip the argument from image
recipes."
Yes, we can do that. But you should also provide a reason why we want to do it there (same argument applies, but weaker, for patch 2).

File: target/linux/sunxi/image/Makefile [hunk 1]
```diff
@@ -32,7 +33,6 @@ endef
 # why \x00\x00\x00\x00 for zImage-initramfs
 define Device/Default
   PROFILES := Default
-  DEVICE_VARS := SUNXI_DTS SUNXI_DTS_DIR SUNXI_UBOOT
```
Comment (adrianschmutzler) [inline target/linux/sunxi/image/Makefile]: Even if this patch is not accepted eventually, we should definitely move DEVICE_VARS out of Device/Default as in this case, as this just makes no sense.

**Comment by mans0n** [comment]
Tweaked the commit messages for patch 2, 3, and 8.

**Comment by mans0n** [comment]
> In most cases check-size is used with IMAGE_SIZE and vice versa.

193 out of 217 for check-size, 193 out of 196 for IMAGE_SIZE in image recipes. So the bond between the pair is quite strong.
Maybe we can add another function like `check-image-size` only to replace `check-size $$$$(IMAGE_SIZE)`.

For patch 8, it's just my observation (based on tests), and I don't really understand the detailed mechanism how Makefile variables and the build scripts work. (I tried but failed.) AFAIK `$$$$()` is also safe (for device recipes), so we might use `$$$$()` for all variables (though I don't like the redundancy).

For meta-device recipes (like Device/Default or Device/tplink-v1), we have three options:
1. Use $$$$().
2. Use `=` instead of `:=`.
3. Don't touch them. Too much headaches.

**Comment by adrianschmutzler** [comment]
> 193 out of 217 for check-size, 193 out of 196 for IMAGE_SIZE in image recipes. So the bond between the pair is quite strong.
> Maybe we can add another function like check-image-size only to replace check-size $$$$(IMAGE_SIZE).

I'm not against this, but currently it is presented as a matter of taste. I actually like your patch 2 with making the parameter just optional. Nevertheless, we should find a reason for it or at least state why our taste is superior ;-)

**Comment by adrianschmutzler** [comment]
> For patch 8, it's just my observation (based on tests), and I don't really understand the detailed mechanism how Makefile variables and the build scripts work. (I tried but failed.) AFAIK `$$$$()` is also safe (for device recipes), so we might use `$$$$()` for all variables (though I don't like the redundancy).
> 
> For meta-device recipes (like Device/Default or Device/tplink-v1), we have three options:
> 
> 1. Use `$$$$()`.
> 2. Use `=` instead of `:=`.
> 3. Don't touch them. Too much headaches.

You have described the problem nicely. I'm always confused with the dollar duplication as well, and I'm not sure whether I want to invest time in sorting this out. So, from my point of view, I'd decide on 1-7 relatively soon and then wait with no. 8 a little longer ...

**Comment by mans0n** [comment]
> I'm not against this, but currently it is presented as a matter of taste.

Yes, this whole PR is mostly a matter of taste (for reducing redundancy). So I added RFC in the title...

> So, from my point of view, I'd decide on 1-7 relatively soon and then wait with no. 8 a little longer ...

Great. We really need more reviews for the patch 8.

**Comment by mans0n** [comment]
Added patch 5 for SUNXI_UBOOT.

File: include/image.mk [hunk 2]
```diff
@@ -381,13 +381,13 @@ define Device/Init
...
   KERNEL:=
   KERNEL_INITRAMFS = $$(KERNEL)
-  KERNEL_SIZE:=
   CMDLINE:=
 
   IMAGES :=
   ARTIFACTS :=
   IMAGE_PREFIX := $(IMG_PREFIX)-$(1)
   IMAGE_NAME = $$(IMAGE_PREFIX)-$$(1)-$$(2)
+  IMAGE_SIZE :=
```
Comment (adrianschmutzler) [inline include/image.mk]: If Device/Init is actually setting per-device global variables, that might be the solution for your problems with DEVICE_TYPE. Just move the global default DEVICE_TYPE here and add it to DEVICE_VARS ...

**Comment by adrianschmutzler** [comment]
I've picked the mentioned patches + SUNXI_UBOOT and pushed them for now.

File: target/linux/bcm63xx/image/bcm63xx.mk [hunk 3]
```diff
@@ -3,14 +3,19 @@
 # BCM33XX/BCM63XX Profiles
 #
 
+DEVICE_VARS += HCS_MAGIC_BYTES HCS_REV_MIN HCS_REV_MAJ
+DEVICE_VARS += BLOCK_SIZE FLASH_MB IMAGE_OFFSET
```
Comment (adrianschmutzler) [inline target/linux/bcm63xx/image/bcm63xx.mk]: I wonder whether we could use BLOCKSIZE from image.mk instead ...

File: include/image.mk (hunk 2 above)
Comment (mans0n) [reply include/image.mk]: Yes, I tried it in my draft patch and it seems to work fine. Thanks for the tip. :)

File: target/linux/bcm63xx/image/bcm63xx.mk (hunk 3 above)
Comment (mans0n) [reply target/linux/bcm63xx/image/bcm63xx.mk]: Hmm maybe, but I don't know enough about CFE to to understand the image recipe. BLOCK_SIZE may have a different meaning than BLOCKSIZE.

**Comment by mans0n** [comment]
Rebased on master.

**Comment by mans0n** [comment]
Rebased due to a66eee63368e and 132ff90f1d54.

**Comment by adrianschmutzler** [comment]
I will build-test and merge the remaining patches except the $$ replace during the next hour.

The magic reordering is based on the commit date, which sometimes is messed up during rebase. The only working solution for me is go through the patches one-by-one with "rebase -i" and update the commit date.

**Comment by adrianschmutzler** [comment]
@mans0n I decided that I won't invest time in reviewing the remaining "apm821xx: replace $$$$() with $$()" patch, as I think the invested time for being sure about the change is not justified by the benefit.

Since nobody else will find that patch in the mess I made out of this PR, I suppose you close this PR and open a new one or send that single remaining patch to the mailing list. Maybe include the explanation from above there as well.

**Comment by mans0n** [comment]
>  I decided that I won't invest time in reviewing the remaining "apm821xx: replace $$$$() with $$()" patch, as I think the invested time for being sure about the change is not justified by the benefit.

I fully understand. Either I don't have much time to invest in this PR these days. :/
I'll close this PR for now, and reopen another one if I manage to get its more complete form.
Thanks for merging the other patches anyway.

**Pull request closed without merging**